src/features/cart/ShoppingCartTest.ts

```diff
@@ -4,171 +4,176 @@ interface Item{
 }
+//Get Single Card Price 
+const singlePrice=(item:Item):number=>{
+    const codeApplied = item.discount.length > 0;
+    let totalPrice=0;
+    if(codeApplied){
+        const codeValue=discountCodes.filter((code)=>item.discount.includes(code.code))
+        console.log(codeValue)
+        const totalDiscount=codeValue.reduce((total,item)=>total+item.value,0)
+        console.log("Total Discount",totalDiscount)
+       totalPrice=(item.unitPrice*item.quantity)-((item.unitPrice*item.quantity*totalDiscount)/100);
+
+    }else{
+        totalPrice = item.quantity * item.unitPrice;
 
-console.log("Single Card Price ",singlePrice);
-
-
-//Get Total Price
-
-const totalPrice=(store:Array<Item>):number=>{
-    const grandTotal=store.reduce((total,item)=>total + item.quantity*item.unitPrice,0);
-    return grandTotal;
-}
-
-console.log("Grand Total ",totalPrice(storeItem));
-
-
-//Show all Item In Cart 
-
-const showAllItems=(cart:Array<Item>):void=>{
-    const allItems=cart.filter((product)=>product.quantity>1)
-    if(allItems){
-        console.log("all Item ")
-    }
-    else{
-        console.log("There are noew order in your card ")
     }
+    return totalPrice;
 }
+const grandTotal=():number=>{
+  // check if any item has discounted price 
+  const haveDiscount = cart.items.filter(item => item.discount.length > 0);
+  let totalDiscountedItemPrice : number = 0;
+  if(haveDiscount){
+    const discountedItemPrice : number[] = [];
+    // get discounted price of these items 
+    haveDiscount.forEach(item => {
+     const price =  singlePrice(item);
+     discountedItemPrice.push(price)
+    })
+    totalDiscountedItemPrice = discountedItemPrice.reduce((total, item) => total + item, 0)
+  }
+  const codeApplied = cart.discount.length > 0;
+  let totalPrice = 0;
+  if (codeApplied) {
+    const codeValue = discountCodes.filter((code) =>
+      cart.discount.includes(code.code)
+    );
+    const totalDiscount = codeValue.reduce(
+      (total, item) => total + item.value,
+      0
+    );
+    const cartPrice = cart.items.reduce((total, item) => {
+      if(item.discount.length === 0){
+        return total + item.quantity * item.unitPrice
+      } else {
+        return total
+      }
+    }, 0);
+    const finalPrice = cartPrice + totalDiscountedItemPrice;
+
+    totalPrice = finalPrice - (finalPrice * totalDiscount / 100);
+  } else {
+    const cartPrice = cart.items.reduce((total, item) => {
+      if(item.discount.length === 0){
+        return total + item.quantity * item.unitPrice
+      } else {
+        return total
+      }
+    }, 0);
+    totalPrice = cartPrice + totalDiscountedItemPrice;
+    
+  }
+  return totalPrice;
+}
+// show all cart items
 
+const getCartItems = (): Item[] => {
+    return cart.items.filter((product) => product.quantity >= 1);
+  };
 
-console.log("All Item Here ",showAllItems(storeItem));
-
-
+export{addToCart,removeToCart,addDiscountCode,getCartItems,grandTotal,singlePrice,cart,discountCodes}
```

Comment:
an empty export is used in code. This may be a bad practice.

Reply:
Ok I got it and Solve now thanks for review this